File: Programação Dinamica/LCS.c
#include <stdio.h>
#define MAX 50

int tamanho(char C[])
{
	int i = 0;
	while(C[i] != '\0')
	{
		i++;
	}
	return i;
}

int main()
{
	char A[MAX], B[MAX];
	int TA, TB, i, j;
	scanf(" %s", A);
	scanf(" %s", B);
	TA = tamanho(A);
	TB = tamanho(B);
	printf("%d %d\n", TA, TB);
	int Matriz[TA + 1][TB + 1];

	for(i = 0; i <= TA; i++)
		Matriz[0][i] = 0;

	for(i = 0; i <= TB; i++)
		Matriz[i][0] = 0;

	for(i = 1; i <= TA; i++){
		for(j = 1; j <= TB; j++){
			if(A[i - 1] == B[j - 1]){
				Matriz[i][j] = Matriz[i - 1][j - 1] + 1;
			}
			else{
				if(Matriz[i - 1][j] > Matriz[i][j - 1]){
					Matriz[i][j] = Matriz[i - 1][j];
				}
				else{
					Matriz[i][j] = Matriz[i][j - 1];
				}
			}
		}
	}

	printf("%d\n", Matriz[TA][TB]);

	return 0;

}

File: Programação Dinamica/lcs.py
#!/usr/bin/python3

def main():
	str1 = input()
	str2 = input()

	matriz = []

	for i in range(len(str1) + 1):
		matriz.append([0 for i in range(len(str2) + 1)])

	print(lcs(matriz, str1, str2))

def lcs(matriz, str1, str2):
	for i in range(1, len(matriz)):
		for j in range(1, len(matriz[0])):
			if str1[i-1] == str2[j-1]:
				matriz[i][j] = matriz[i-1][j-1] + 1
			else:
				matriz[i][j] = max(matriz[i-1][j], matriz[i][j-1])
	return matriz[len(str1)][len(str2)]

if __name__ == '__main__':
	main()